lib/TabTracker.js

```diff
@@ -0,0 +1,89 @@
+const tabs = require("sdk/tabs");
+const ss = require("sdk/simple-storage");
+
+const NEWTAB_URL = "resource://activity-streams/data/content/activity-streams.html#/";
+
+const TabTracker = {
+  _openTabs: [],
+
+  init: function() {
+    this.onOpen = this.onOpen.bind(this);
+    tabs.on("open", this.onOpen);
+    if (!ss.storage.tabData) {
+      ss.storage.tabData = {};
+    }
+  },
+
+  uninit: function() {
+    for (let i in this._openTabs) {
+      let tab = this._openTabs[i];
+      tab.removeListener("pageshow", this.logShow);
+      tab.removeListener("activate", this.logActivate);
+      tab.removeListener("deactivate", this.logDeactivate);
+      tab.removeListener("close", this.logClose);
+    }
+    tabs.removeListener("open", TabTracker.onOpen);
+  },
+
+  navigateAwayFromNewtab: function(tab) {
+    if (!ss.storage.tabData[tab.id]) {
+      // We're navigating away from a newtab that didn't even load yet.
+      // Let's say it's been active for 0 seconds.
+      ss.storage.tabData[tab.id] = {activations: [{startTime: Date.now(), totalTime: 0}]};
+      return;
+    }
+    let activations = ss.storage.tabData[tab.id].activations;
+    let lastElement = activations[activations.length - 1];
+    if (lastElement.startTime && !lastElement.totalTime) {
+      lastElement.totalTime = (Date.now() - lastElement.startTime) / 1000;
+    }
```

Comment:
we might want to `Cu.reportError` or `console.warn` in case `totalTime` is not falsey. It means this callback has been called multiple times